カレンダープログラムを作成

naoki00413 opened the pull request.

Commit:
fizzbuzz課題

naoki00413 committed:
指摘箇所を修正

naoki00413 committed:
fizzbuzz.rbファイルを削除

ファイルが2個になってしまった為、余分な方を削除

naoki00413 committed:
指摘箇所を修正

ファイル名を'fizz.rb'から'fizzbuzz.rb'に変更
指摘にあった演算子にスペースを追加

naoki00413 committed:
Merge branch 'fizzbuzz' of https://github.com/naoki00413/ruby-practices into fizzbuzz

naoki00413 committed:
カレンダー課題の提出

カレンダー課題のコードです

JunichiIto commented on 01.fizzbuzz/fizzbuzz.rb:
```diff
@@ -0,0 +1,12 @@
+(1..20).each do |i|
```
この課題はカレンダーの課題なので、fizz buzzが含まれているのはおかしいです。
今回はこのままで良いですが、次回以降は次のような手順でブランチを作ってください。

前提条件
- fizz buzzはまだ合格していないが、カレンダーに着手したい
- ローカルの現在のブランチはfizzbuzzz

手順
- ローカルのブランチをmainにする
- そこからカレンダー用のブランチを作成する
- カレンダープログラムを作る
- git hubにカレンダー用ブランチをpushする
- カレンダーのPRを作る

以降、fizzbuzzを修正したいときはfizzbuzzブランチを、カレンダーを修正したいときはカレンダー用のブランチに切り替えてコードを修正してください。

JunichiIto commented on 02.calendar/calendar.rb:
```diff
@@ -0,0 +1,23 @@
+require "date"
+require 'optparse'
+# 年月の初期設定
```
コメントはWHYだけを書きましょう。

[【動画付き・初心者向け】コメントにはWHYを書こう 〜不要なコメントと必要なコメントの違いについて〜 \- Qiita](https://qiita.com/jnchito/items/f0d90af4ed44b7484103)

JunichiIto commented on 02.calendar/calendar.rb:
```diff
@@ -0,0 +1,23 @@
+year = params.values[0].to_i
+month = params.values[1].to_i
+# 年月の出力
+year_month = month.to_s + "月" .center(2)+ year.to_s
```
`"#{month}月#{year}"`のように、`#{}`を使うとRubyらしくなります。
あと、空白を2つ入れたいのであればわざわざ`center(2)`を使う必要はないかなと思いました。

JunichiIto commented on 02.calendar/calendar.rb:
```diff
@@ -0,0 +1,23 @@
+# 年月の出力
+year_month = month.to_s + "月" .center(2)+ year.to_s
+# 出力して年月を中央寄せにする
+puts year_month.center(18)
```
本家カレンダーと微妙にずれてるようです。

![isnu9iajy0cv4chv5ueqi98qqs2v](https://user-images.githubusercontent.com/1148320/218894243-c3d03036-47d3-428f-8326-585163ec90f4.jpg)

JunichiIto commented on 02.calendar/calendar.rb:
```diff
@@ -0,0 +1,23 @@
+# 曜日の出力
+puts "日 月 火 水 木 金 土"
+# 日付の出力
+first_wday = Date.new(year,month,1).wday
```
```suggestion
first_wday = Date.new(year, month, 1).wday
```

のようにカンマの後ろにスペースを入れましょう

JunichiIto commented on 02.calendar/calendar.rb:
```diff
@@ -0,0 +1,23 @@
+     if (first_wday + day) % 7 == 0 || day == last_day
+      puts "\n"
+     end
+    }
```
18〜23行目、インデントが崩れてるので正しく修正してください。

JunichiIto commented on 02.calendar/calendar.rb:
```diff
@@ -0,0 +1,23 @@
+(1..last_day).each { |day|
+    printf("%2d ", day)
+     if (first_wday + day) % 7 == 0 || day == last_day
+      puts "\n"
```
puts自体が改行を伴う出力なので（`"\n"`はおそらく無視されている）、`print "\n"`の方が違和感がないように思います。

JunichiIto reviewed (commented):
コメントしました！

naoki00413 committed:
カレンダープログラムの指摘点を修正

修正点 不要なコメントを削除, 出力のコードを少し変更
本家カレンダーと表記が合うように修正, カンマにスペースを追加
崩れてたインデントを修正, 改行の出力を違和感ないように修正

JunichiIto commented on 02.calendar/calendar.rb:
```diff
@@ -0,0 +1,22 @@
+  printf("%2d ", day)
+    if (first_wday + day) % 7 == 0 || day == last_day
+    print "\n"
+    end
```
19-21行目のインデントがおかしいので修正してください〜。

JunichiIto commented on 02.calendar/calendar.rb:
```diff
@@ -0,0 +1,22 @@
+last_day = Date.new(year, month, -1).day
+print "   " * first_wday
+
+(1..last_day).each { |day|
```
複数行のブロックは`{}`よりも`do/end`が使われることが多いです。

あと、ループ処理は整数値の1から31をループさせるより、Dateオブジェクトをループさせる方がいろいろコードをスッキリさせやすくなります。

```ruby
first_date = Date.new(year, month, 1)
last_date = Date.new(year, month, -1)
(first_date..last_date).each do |date|
  # ...
```

JunichiIto reviewed (commented):
コメントしました！

naoki00413 committed:
カレンダープログラム 指摘点のブロックを{}を変更

修正点：指摘点の{}からdo/endに修正、19-21行目のインデントを再修正

naoki00413 committed:
ループ処理をDateオブジェクトに変更

修正点:ループ処理を1~31からDateオブジェクトに変更

JunichiIto commented on 02.calendar/calendar.rb:
```diff
@@ -10,13 +10,14 @@
-first_wday = Date.new(year, month, 1).wday
-last_day = Date.new(year, month, -1).day
-print "   " * first_wday
+first_date = Date.new(year, month, 1).day
```
dayはDateオブジェクトではなく、整数値（Integer）です。

https://docs.ruby-lang.org/ja/latest/method/Date/i/day.html

なので、僕としては以下のように書く方が望ましいと考えていますー。

```suggestion
first_date = Date.new(year, month, 1)
```

JunichiIto commented on 02.calendar/calendar.rb:
```diff
@@ -10,13 +10,14 @@
-    if (first_wday + day) % 7 == 0 || day == last_day
+(first_date..last_date).each do |date|
+  printf("%2d ", date)
+  if (first_day + date) % 7 == 0
```
Dateオブジェクトを活用すれば、この行もこんなふうに書けるはずです。

```suggestion
  if date.saturday?
```

JunichiIto reviewed (commented):
コメントしました！

naoki00413 committed:
日付の出力の仕方を変更

Dateオブジェクトを活用した書き方に変更
それに伴い`first_day`を削除

JunichiIto reviewed (approved):
確認しました。これでOKです〜 🙆‍♂️
